indexer-app/src/cmr/indexer/data/concepts/keyword_util.clj

```diff
@@ -20,5 +20,152 @@
   [field-values]
   (->> field-values
        (mapcat prepare-keyword-field)
-       set
-       (string/join " ")))
+       (keep not-empty)
+       (apply sorted-set)
+       (string/join \space)))
+
+(defn contact-group->keywords
+  "Converts a contact group into a vector of terms for keyword searches."
+  [contact-group]
+  (let [{group-name :GroupName
+         roles :Roles} contact-group]
+    (concat [group-name]
+            roles)))
+
+(defn contact-person->keywords
+  "Converts a compound field into a vector of terms for keyword searches."
+  [data]
+  (let [{first-name :FirstName
+         last-name :LastName
+         roles :Roles} data]
+    (concat [first-name last-name]
+            roles)))
+
+(defn names->keywords
+  "Converts a compound field into a vector of terms for keyword searches."
+  [data]
+  (let [{long-name :LongName
+         short-name :ShortName} data]
+    [long-name
+     short-name]))
+
+(defn platform->keywords
+  "Converts a compound field into a vector of terms for keyword searches."
+  [data]
+  (let [{instruments :Instruments} data]
+    (concat (names->keywords data)
+            (mapcat names->keywords instruments))))
+
+(defn related-url->keywords
+  "Converts a compound field into a vector of terms for keyword searches."
+  [data]
+  (let [{description :Description
+         subtype :Subtype
+         type :Type
+         url :URL
+         url-content-type :URLContentType} data]
+    [description
+     subtype
+     type
+     url
+     url-content-type]))
+
+(defn science-keyword->keywords
+  "Converts a science keyword into a vector of terms for keyword searches."
+  [science-keyword]
+  (let [{category :Category
+         detailed-variable :DetailedVariable
+         term :Term
+         topic :Topic
+         variable-level-1 :VariableLevel1
+         variable-level-2 :VariableLevel2
+         variable-level-3 :VariableLevel3} science-keyword]
+    [category
+     detailed-variable
+     term
+     topic
+     variable-level-1
+     variable-level-2
+     variable-level-3]))
+
+(defn service-keyword->keywords
+  "Converts a service keyword into a vector of terms for keyword searches."
+  [service-keyword]
+  (let [{service-category :ServiceCategory
+         service-specific-term :ServiceSpecificTerm
+         service-term :ServiceTerm
+         service-topic :ServiceTopic} service-keyword]
+    [service-category
+     service-specific-term
+     service-term
+     service-topic]))
+
+(defn service-organization->keywords
+  "Converts a service keyword into a vector of terms for keyword searches."
+  [service-organization]
+  (let [{roles :Roles
+         service-contact-persons :ContactPersons} service-organization]
+    (concat (names->keywords service-organization)
+            (mapcat contact-person->keywords service-contact-persons)
+            roles)))
+
+(def fields->fn-mapper
```

Comment:
I feel uneasy with this function grouping fields from both variable and service concepts. I see the following issues:

1/ It makes the assumption that the fields will have the same structure for different UMM models. It is true today, but may not be true in the future.
2/ This puts different fields from different concepts in a single map. You have to look at the variables and services code to see which `schema-keys` are actually used for each concept type. 
3/ Collection is not using the `concept-keys->keyword-text` approach to populate its keyword value. This makes the approach overly complicated for just variable and service concepts.

I think what I am saying is that it would be easier to follow if we just use the same approach Collection is using to just make a separate function to generate keyword values for Variable and Service concept. That way all the keywords fields for different concepts are grouped together and created in the same way and it would be easier to maintain.

Reply:
Lemme take care of some of the smaller review items, and then I'll give this some thought and an update then ...

Reply:
> 2/ This puts different fields from different concepts in a single map. You have to look at the variables and services code to see which schema-keys are actually used for each concept type.

Yeah, the part of this that I don't like is really just more visual/aestheic: when you look at the map, there's no obvious self-documentation. I objected but kept going, seeing this as an interim approach.  

Hrm, I just thought of a brain-dead way of making this more clear without having to change anything, really ... be right back!

Reply:
Okay, fix for your comment #2 looks good. Will be in next push.

Reply:
> 3/ Collection is not using the concept-keys->keyword-text approach to populate its keyword value. This makes the approach overly complicated for just variable and service concepts.

Couple of points here:

1) This PR does update Collection to use `science-keyword->keywords` as well as `fields->fn-mapper` for `:variable-name` and  `:measurement` -- which are the only things with shared/overlapping functionality where Collections were already using something in `cmr.indexer.data.concepts.keyword-util`.

2) It may seem overly complicated to you while reading through the PR, but this was not my original solution ... the original solution I had in place duplicated a ton of logic and made for multiple, messing `let` blocks spread throughout the code and tests. It was that which pushed me to  come up with a better, more reusable solution.

3) `#'cmr.indexer.data.concepts.collection.keyword/create-keywords-field` is a mess and should be fixed (I'll create a ticket); it shouldn't be used to measure any current implementation.

4) "Easier to follow" is a fairly subjective term which can conflate familiarity with good coding practices. We should make decisions based upon the latter ... Note that in `cmr.indexer.data.concepts.keyword-util`, each `X->Y` function has a docstring, as does the mapper data structure `def` and the general `concept->keyword` functions at the end. One can easily read through each of these to find out what is going on. Alternatively, if you look at the `let` block in `cmr.indexer.data.concepts.collection.keyword/create-keywords-field`, you have none of that :-( there are tons of `map`s, `mapcat`s, and `keep`s with no guide to the future developer of what's really going on ... and all that logic is trapped in the `let` block :-(

Reply:
I feel the `let` block is actually easier to follow, but the individual function approach is more modular and easier to test. So at this point since you already did the work to refactor, I am OK with you just merge what you have.